src/journal/JournalPlayer.cc

```diff
@@ -800,4 +828,43 @@ void JournalPlayer::notify_complete(int r) {
     m_replay_handler), r);
 }
 
+void JournalPlayer::handle_cache_rebalanced(uint64_t new_cache_bytes) {
+  Mutex::Locker locker(m_lock);
+
+  auto splay_width = m_journal_metadata->get_splay_width();
+  m_max_fetch_bytes = p2align<uint64_t>(new_cache_bytes / splay_width, 4096);
+
+  ldout(m_cct, 10) << __func__ << ": new_cache_bytes=" << new_cache_bytes
+                   << ", max_fetch_bytes=" << m_max_fetch_bytes << dendl;
+
+  uint64_t min_bytes = 32768;
+
+  if (m_state == STATE_WAITCACHE) {
+    if (m_max_fetch_bytes >= min_bytes) {
+      m_state = STATE_INIT;
+      auto ctx = new FunctionContext(
+        [this](int r) {
+          prefetch();
+        });
+      m_journal_metadata->queue(ctx, 0);
+      return;
+    }
+  } else {
+    min_bytes /= 2;
```

Comment:
What's the purpose behind this? Just so it creates a new minimum when not prefetching?

Reply:
The idea is that when we are short on memory it is preferable not to start a new JournalPlayer instead of shutting down the already running one. So on the start the minimum required is two times higher than when it is already running.

Still, I worry a little about a potential effect, when many images are replaying, no new ones are started and then by some reason the limit is reached. Then all replayers will be stopped simultaneously (because the balancer will reserve equally not enough space for them all). After this they all will be restarted and only part of them (the first ones) will be able to start, so it will functioning, but this shutdown looks like undesirable.

Reply:
Is there a way that we can have it fail-fast? I agree it would be awkward if suddenly previously running replication was stopped just because a new image was added. That's kinda why I was wondering about adding the concept of reserving a minimum size from the cache for each ```JournalPlayer``` upon start-up. A player might be given more than that minimum amount if there is space but it will never be given less (maybe ```rbd_mirror_journal_max_fetch_bytes``` just becomes ```rbd_mirror_journal_min_fetch_bytes```?).

Reply:
To resolve this potential issue we could add some randomization:
- e.g. min_bytes instead of dividing by 2 could use some randomized factor so only part of the images are likely to reach `min_bytes` and would be stopped.
- or if min_bytes is reached we stop only with a probability 0.5 or something like this.

Reply:
What about moving the ```m_cache_manager_handler``` call to ```prefetch``` and allow ```register_cache``` to return ```-ENOMEM``` if there is no more space for even the min fetch bytes? This error would then get bubbled up to the replay handler.